### DISCUSSION TITLE
Credentials Provider Refresh Token Rotation - jwt callback token doesn't update after initial sign in.

### DISCUSSION THREAD

**Discussion by jan-grasewicz** [nextauthjs/next-auth#2850] [Help]
### Question 💬

Hi.
I'm trying to setup refresh token rotation using Credentials Provider.
Based on this example: https://next-auth.js.org/v3/tutorials/refresh-token-rotation

Unfortunate jwt callback saves returned token only on initial sign in and when getSession is called client-side, not  when getSession is called in getServerSideProps function.

So my access token gets refreshed as long as I do client-side requests, but if I fetch all my data server-side my session will expire.

Am I doing something wrong in my config, or is it a bug?

### How to reproduce ☕️

[...nextauth].js
```js
export default NextAuth({
  providers: [
    CredentialsProvider({
      name: 'Sign in with email and password',
      credentials: {
        email: {
          label: 'Email',
          type: 'email',
          placeholder: 'email@domain.com',
        },
        password: { label: 'Password', type: 'password' },
      },
      async authorize(credentials) {
          const res = await fetch(
            process.env.NEXTAUTH_API_URL + '/auth/sign-in',
            {
              method: 'POST',
              body: JSON.stringify(credentials),
              headers: { 'Content-Type': 'application/json' },
            },
          );

          const user = await res.json();

          if (res.ok && user) {
            const { accessToken, accessTokenExpirationTime, refreshToken } =
              user.metadata;

            return {
              ...user.result,
              accessToken,
              accessTokenExpires: Date.now() + accessTokenExpirationTime,
              refreshToken,
            };
          }

          return null;
      },
    }),
  ],
  pages: {
    signIn: '/auth/login',
  },
  callbacks: {
    async jwt(token, user, account) {
      // Initial sign in
      if (account && user) {
        const { accessToken, accessTokenExpires, refreshToken, ...userData } =
          user;
        return {
          accessToken,
          accessTokenExpires,
          refreshToken,
          user: userData,
        };
      }

      if (Date.now() < token.accessTokenExpires) {
        return token;
      }

      const updatedToken = await refreshAccessToken(token);

      return updatedToken;
    },
    async session(session, token) {
      session.user = token.user;
      session.accessToken = token.accessToken;
      session.error = token.error;

      return session;
    },
  },
});

async function refreshAccessToken(token) {
  try {
    const cookie = `Authentication=${token.accessToken}; Refresh=${token.refreshToken}; `;
    const response = await fetch(
      process.env.NEXTAUTH_API_URL + '/auth/refresh',
      { method: 'POST', headers: { cookie } },
    );

    const refreshedTokens = await response.json();

    if (!response.ok) {
      throw refreshedTokens;
    }

    const { accessToken, accessTokenExpirationTime, refreshToken } =
      refreshedTokens.metadata;

    const updatedToken = {
      ...token,
      accessToken,
      accessTokenExpires: Date.now() + accessTokenExpirationTime,
      refreshToken: refreshToken ?? token.refreshToken, // Fall back to old refresh token
    };

    return updatedToken;
  } catch (error) {
    console.error('RefreshAccessTokenError', error);
    return {
      ...token,
      error: 'RefreshAccessTokenError',
    };
  }
}
```

"refreshAccessToken" always returns new tokens, but on subsequent calls "jwt(token)" returns them only if refresh was the result of "getSession" called client-side, not in getServerSideProps

I'm using v3, but tried v4 with no success.

### Contributing 🙌🏽

No, I am afraid I cannot help regarding this

**Comment by mackankowski** [comment]
The following topic looks similar and it doesn't seem to be any solution since May :( 

https://github.com/nextauthjs/next-auth/discussions/2003

**Comment by balazsorban44** [comment]
`getSession` works exactly the same for backend and front-end usage (something which I actually want to change because it is a performance bottleneck #1535), so it does the same in both places.

When used server-side though, you should pass `req` to `getSession` as documented here:
https://next-auth.js.org/getting-started/client#server-side-example

In the future, we likely will have a `getServerSession` or similar method, to avoid confusion.

**Reply by jan-grasewicz** [reply]
I'm using getSession in getServerSideProps like so:
```
export const getServerSideProps = async (ctx) => {
    const session = await getSession(ctx);
    
    // auth redirect logic here
    
    return {
        props: {}
    }
}
```
and the logs I get are different from getSession used client-side

```
🚀 JWT initial sign in {
  token: {
    name: undefined,
    email: 'user@domain.com',
    picture: undefined,
    sub: 'c43d75ad-9597-41be-89ef-f3a53c27708d'
  },
  initialAccessToken: 'AAA'
}
>> next-auth session.js🚀 decodedJwt: AAA
🚀 JWT valid token {accessToken: 'AAA'}
SESSION 🚀 {sessionAccessToken: 'AAA'}
>> next-auth session.js🚀 decodedJwt: AAA
🚀 JWT valid token {accessToken: 'AAA'}
SESSION 🚀 {sessionAccessToken: 'AAA'}
>> next-auth session.js🚀 decodedJwt: AAA
🚀 JWT valid token {accessToken: 'AAA'}
SESSION 🚀 {sessionAccessToken: 'AAA'}
//// client side getSession
>> next-auth session.js🚀 decodedJwt: AAA
=======REFRESH==========
🚀 JWT token expired {oldAccessToken: 'AAA', updatedAccessToken: 'BBB'}
SESSION 🚀 {sessionAccessToken: 'BBB'}
>> next-auth session.js🚀 decodedJwt: BBB // IMPORTANT lib returns new(correct BBB) token after refresh
🚀 JWT valid token {accessToken: 'BBB'}
SESSION 🚀 {sessionAccessToken: 'BBB'}
//// server side getSession
>> next-auth session.js🚀 decodedJwt: BBB
=======REFRESH==========
🚀 JWT token expired {oldAccessToken: 'BBB',updatedAccessToken: 'CCC'}
SESSION 🚀 {sessionAccessToken: 'CCC'}
>> next-auth session.js🚀 decodedJwt: BBB // IMPORTANT lib returns old(expect CCC) token after refresh
=======REFRESH==========
RefreshAccessTokenError {errorAccessToken: 'BBB'} {
  statusCode: 401,
  message: 'No token match',
  response: { statusCode: 401, message: 'No token match', error: 'Unauthorized' },
  timestamp: '2021-09-28T08:41:00.845Z',
  path: '/api/auth/refresh'
}
🚀 JWT token expired {oldAccessToken: 'BBB', updatedAccessToken: 'BBB'}
SESSION 🚀 {sessionAccessToken: 'BBB'}
```

**Reply by balazsorban44** [reply]
I'm really struggling to understand the flow of your logs. :sweat_smile: 

what is session.js?

A runnable/cloneable reproduction would be more helpful, actually.

**Reply by jan-grasewicz** [reply]
sorry session.js is here https://github.com/nextauthjs/next-auth/blob/main/src/server/routes/session.js ;)

I'll clean up the the log a bit, but creating a running reproduction is a bit tricky, as I would have to find some placeholder backend or mock it.

Thanks for the interest :)

**Reply by jan-grasewicz** [reply]
JWT and SESSION correspond to callbacks
```
JWT 🚀 initial sign in accessToken: 'AAA'
JWT 🚀 valid token accessToken: 'AAA'
SESSION 🚀  accessToken: 'AAA'

====>client side getSession<====

>> refreshAccessToken in [...nextauth].js <<
JWT 🚀 token expired / oldAccessToken: 'AAA' updatedAccessToken: 'BBB'
SESSION 🚀 accessToken: 'BBB'

!!!!! IMPORTANT lib returns new (BBB-correct) token after refresh

JWT 🚀 valid token / accessToken: 'BBB'
SESSION 🚀 accessToken: 'BBB'

====>server side getSession<====

>> refreshAccessToken in [...nextauth].js <<
JWT 🚀 token expired / oldAccessToken: 'BBB' updatedAccessToken: 'CCC'
SESSION 🚀 accessToken: 'CCC'

!!!!! IMPORTANT lib returns old(BBB-incorrect) token after refresh 

>> refreshAccessToken in [...nextauth].js <<
RefreshAccessTokenError  accessToken: 'BBB'
JWT 🚀 token expired / oldAccessToken: 'BBB' updatedAccessToken: 'BBB'
SESSION 🚀 accessToken: 'BBB'
```
getSession starts a refreshAccessToken
When called from client-side, updated token, returned from jwt callback, gets saved. In future is available in callbacks: { async jwt(token) }.
When called from getServerSideProps updated token is returned from jwt and gets passed to session callback, but after that jwt callback returns old access token.

**Reply by balazsorban44** [reply]
A placeholder backend could be just some hardcoded objects. I suspect the problem doesn't lie in generating the correct values.

**Reply by jan-grasewicz** [reply]
I stand corrected. :)
Here you go: https://codesandbox.io/s/cool-marco-mzy3t?file=/pages/api/auth/%5B...nextauth%5D.js

**Reply by mackankowski** [reply]
I have a feeling that it could be a bug. I would play more with the library to find out if everything is passed correctly inside.

**Reply by balazsorban44** [reply]
Here is the relevant logic:
https://github.com/nextauthjs/next-auth/blob/e8a58a01b6b8c456268771f9b382fa300ccab8ac/src/server/routes/session.js#L22-L68

This will run for `getSession`, no matter if it's invoked on the client or the server.

You can see that we run both `jwt` and `session` callbacks after another, and update the cookie with the new values.

**Reply by jan-grasewicz** [reply]
Do I understand correctly that `newEncodedJwt` stored in cookie, is the only place that we keep our tokens?
When called from `getServerSideProps` the header `Set-Cookie` from `getSession response`, won't change client side cookies. Am I right?

**Reply by balazsorban44** [reply]
if you pass the whole context to getSession, it should be. you can set headers/cookies from getServerSideProps

**Reply by balazsorban44** [reply]
although using getSession in server-side has been problematic for other reasons. #1535

I want to introduce a new method called `getServerSession` that would work much better.

**Reply by jan-grasewicz** [reply]
> if you pass the whole context to getSession, it should be. you can set headers/cookies from getServerSideProps

I'm passing whole context in my code sandbox, but new cookies from getSession don't end up in the browser.

**Reply by balazsorban44** [reply]
@jan-grasewicz I finally had the time to get around this, and I can acknowledge the issue! Indeed, calling `getSession` doesn't (and by bad implementation **cannot**) update cookies!

I realized this while working on #2857. So the best solution I have right now is https://github.com/nextauthjs/next-auth/discussions/2850#discussioncomment-1409649

I managed to implement a `getServerSession` method which is available in this experimental release:
https://github.com/nextauthjs/next-auth/pull/2857#issuecomment-930139134

I confirmed that this will indeed update cookies correctly by this simple callback:
```js
callbacks: {
  async session({ session }) {
    session.random = Math.random()
    return session
  },
},
```

Since this creates new data every time, the cookie is supposed to be changing as well. Using `getSession` has no effect on the cookie, while `getServerSession` works as expected!

Considering the implementation of `getSession` it is not really surprising to be honest, it never actually uses the `res` to set cookies, even if you pass the whole `context`.

So for now, you can try out the above-mentioned experimental release (which is based on v4 https://next-auth.js.org/getting-started/upgrade-v4).

Documentation comes later, but have a look at these files to get the idea:

https://github.com/nextauthjs/next-auth/blob/9aba57396c6776642ed743b60a266a29385652ac/app/pages/api/auth/%5B...nextauth%5D.ts
![image](https://user-images.githubusercontent.com/18369201/135545156-8c1a19f5-57ba-4d28-b400-f1b2c87e8183.png)


https://github.com/nextauthjs/next-auth/blob/9aba57396c6776642ed743b60a266a29385652ac/app/pages/protected-ssr.js
![image](https://user-images.githubusercontent.com/18369201/135545103-f10af791-4511-433e-8571-23f34b82eb2f.png)

**Reply by jan-grasewicz** [reply]
> 
> Considering the implementation of `getSession` it is not really surprising to be honest, it never actually uses the `res` to set cookies, even if you pass the whole `context`.
> 
That's what I checked after my last reply.

Thank you for confirming the problem I expected, and providing currently available solution.
Is there any timeline when `getServerSession` might end up in beta release at least?

**Reply by balazsorban44** [reply]
no, there is no ETA

**Reply by magicspon** [reply]
Any movement on this?

**Reply by lws803** [reply]
Running into the same issue now with rotating refresh tokens and calling `getSession` within `getServerSideProps`. Is there a resolution for this? 😢

**Reply by dariolongo** [reply]
Hi all,
is there any update about **unstable_getServerSession**? when there will be an official getServerSession?
Thanks
